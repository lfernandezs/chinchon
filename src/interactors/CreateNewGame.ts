import Game from '../entities/Game/Game';
import { IGameProps } from '../entities/Game/types';
import GameStore from '../stores/GameStore';

interface ICreateNewGameProps {
	gameStore: GameStore;
}

export default class CreateNewGame {
	private _gameStore: ICreateNewGameProps['gameStore'];

	constructor({ gameStore }: ICreateNewGameProps) {
		this._gameStore = gameStore;
	}

	execute({ id, players }: IGameProps) {
		const game = new Game({ id, players });
		this._gameStore.add(game);
	}
}
