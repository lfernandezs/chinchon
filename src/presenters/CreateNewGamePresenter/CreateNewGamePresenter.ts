import { makeAutoObservable } from 'mobx';
import { MAX_NUMBER_OF_PLAYERS } from '../../constants/constants';
import iCreateNewGamePresenterProps from './CreateNewGamePresenter.types';

class CreateNewGamePresenter {
	navigation: iCreateNewGamePresenterProps['navigation'];
	_numberOfPlayers: number;

	constructor({ navigation }: iCreateNewGamePresenterProps) {
		this.navigation = navigation;
		this._numberOfPlayers = 2;
		makeAutoObservable(this);
	}

	get numberOfPlayers() {
		return this._numberOfPlayers;
	}

	get plusButtonDisabled() {
		return this._numberOfPlayers >= MAX_NUMBER_OF_PLAYERS;
	}

	onPlusButtonPress = () => {
		this._numberOfPlayers = Math.min(
			this._numberOfPlayers + 1,
			MAX_NUMBER_OF_PLAYERS
		);
	};

	focusInput(key: number) {
		return key > 1 && key === this._numberOfPlayers - 1;
	}
}

export default CreateNewGamePresenter;
