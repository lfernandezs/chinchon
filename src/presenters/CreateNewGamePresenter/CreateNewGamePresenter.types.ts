interface iCreateNewGamePresenterProps {
	navigation: any;
}

export default iCreateNewGamePresenterProps;
