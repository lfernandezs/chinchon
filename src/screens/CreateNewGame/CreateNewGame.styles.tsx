import { StyleSheet, ViewStyle } from 'react-native';

const plusButton = {
	width: 40,
	height: 40,
	borderRadius: 20,
	alignItems: 'center',
	justifyContent: 'center',
} as ViewStyle;

const styles = StyleSheet.create({
	container: {
		flex: 1,
		alignItems: 'center',
		justifyContent: 'center',
	},
	title: {
		fontSize: 40,
		fontWeight: 'bold',
		marginBottom: 20,
	},

	input: {
		width: '80%',
		height: 40,
		borderColor: 'gray',
		borderWidth: 1,
		marginBottom: 20,
		padding: 10,
	},

	plusButtonEnabled: {
		...plusButton,
		backgroundColor: 'black',
	},

	plusButtonDisabled: {
		...plusButton,
		backgroundColor: 'gray',
	},

	plus: {
		fontSize: 20,
		color: 'white',
	},
});

export default styles;
