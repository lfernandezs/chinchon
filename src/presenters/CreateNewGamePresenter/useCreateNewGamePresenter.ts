import { useNavigation } from '@react-navigation/native';
import { useState } from 'react';
import CreateNewGamePresenter from './CreateNewGamePresenter';

const useCreateNewGamePresenter = () => {
	const navigation = useNavigation();
	const createPresenter = () => {
		return new CreateNewGamePresenter({ navigation });
	};

	const [presenter] = useState(createPresenter);
	return presenter;
};

export default useCreateNewGamePresenter;
