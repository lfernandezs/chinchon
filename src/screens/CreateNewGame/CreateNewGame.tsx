import React from 'react';
import { Text, TextInput, TouchableOpacity, View } from 'react-native';
import useCreateNewGamePresenter from '../../presenters/CreateNewGamePresenter/useCreateNewGamePresenter';
import styles from './CreateNewGame.styles';
import { observer } from 'mobx-react-lite';
import BaseView from '../../components/BaseView/BaseView';

const CreateNewGame = () => {
	const presenter = useCreateNewGamePresenter();
	return (
		<BaseView scrollRef={presenter.scrollRef}>
			<View style={styles.container}>
				<Text testID={'title'} style={styles.title}>
					Jugadores
				</Text>
				{[...Array(presenter.numberOfPlayers)].map((value, key) => (
					<TextInput
						testID={`player-input-${key}`}
						key={key}
						style={styles.input}
						placeholder={`Jugador ${key + 1}`}
						autoFocus={presenter.focusInput(key)}
					/>
				))}
				<TouchableOpacity
					testID={'plus-button'}
					style={
						!presenter.plusButtonDisabled
							? styles.plusButtonEnabled
							: styles.plusButtonDisabled
					}
					onPress={presenter.onPlusButtonPress}
					disabled={presenter.plusButtonDisabled}
				>
					<Text testID={'plus-button-text'} style={styles.plus}>
						+
					</Text>
				</TouchableOpacity>

				<TouchableOpacity
					testID={'create-new-game-button'}
					style={styles.createNewGameButton}
					onPress={presenter.onCreateNewGameButtonPress}
				>
					<Text
						testID={'create-new-game-button-text'}
						style={styles.createNewGameButtonText}
					>
						Crear partida
					</Text>
				</TouchableOpacity>
			</View>
		</BaseView>
	);
};

export default observer(CreateNewGame);
