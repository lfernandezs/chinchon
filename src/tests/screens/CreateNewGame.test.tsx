import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import * as useCreateNewGamePresenter from '../../presenters/CreateNewGamePresenter/useCreateNewGamePresenter';
import CreateNewGame from '../../screens/CreateNewGame/CreateNewGame';
import mockUsePresenter from '../mocks/mockUsePresenter';
import '@testing-library/jest-native/extend-expect';

const presenter = {
	numberOfPlayers: 2,
	plusButtonDisabled: false,
	onPlusButtonPress: jest.fn(),
	focusInput: jest.fn(),
};

mockUsePresenter(useCreateNewGamePresenter, presenter);

describe('CreateNewGame', () => {
	it('renders correctly', () => {
		const screen = render(<CreateNewGame />);

		expect(screen).toMatchSnapshot();
	});

	describe('when the screen is rendered', () => {
		it('should render the title', () => {
			const { getByTestId } = render(<CreateNewGame />);

			expect(getByTestId('title')).toBeTruthy();
			expect(getByTestId('title')).toHaveTextContent('Jugadores');
		});

		it('renders the input for each player', () => {
			const { getAllByTestId } = render(<CreateNewGame />);

			expect(getAllByTestId('player-input', { exact: false })).toHaveLength(
				presenter.numberOfPlayers
			);
		});

		it('renders the plus button', () => {
			const { getByTestId } = render(<CreateNewGame />);

			expect(getByTestId('plus-button')).toBeTruthy();
		});

		it('renders the plus button text', () => {
			const { getByTestId } = render(<CreateNewGame />);

			expect(getByTestId('plus-button-text')).toBeTruthy();
			expect(getByTestId('plus-button-text')).toHaveTextContent('+');
		});
	});

	describe('when the plus button is pressed', () => {
		it('calls the presenter function', () => {
			const { getByTestId } = render(<CreateNewGame />);
			const button = getByTestId('plus-button');
			fireEvent.press(button);
			expect(presenter.onPlusButtonPress).toHaveBeenCalled();
		});
	});
});
