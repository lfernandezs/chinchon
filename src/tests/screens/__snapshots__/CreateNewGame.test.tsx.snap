// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewGame renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingVertical": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 40,
            "fontWeight": "bold",
            "marginBottom": 20,
          }
        }
        testID="title"
      >
        Jugadores
      </Text>
      <TextInput
        placeholder="Jugador 1"
        style={
          Object {
            "borderColor": "gray",
            "borderWidth": 1,
            "height": 40,
            "marginBottom": 20,
            "padding": 10,
            "width": "80%",
          }
        }
        testID="player-input-0"
      />
      <TextInput
        placeholder="Jugador 2"
        style={
          Object {
            "borderColor": "gray",
            "borderWidth": 1,
            "height": 40,
            "marginBottom": 20,
            "padding": 10,
            "width": "80%",
          }
        }
        testID="player-input-1"
      />
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "black",
            "borderRadius": 20,
            "height": 40,
            "justifyContent": "center",
            "opacity": 1,
            "width": 40,
          }
        }
        testID="plus-button"
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 20,
            }
          }
          testID="plus-button-text"
        >
          +
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "black",
            "height": 40,
            "justifyContent": "center",
            "marginTop": 20,
            "opacity": 1,
            "width": "80%",
          }
        }
        testID="create-new-game-button"
      >
        <Text
          style={
            Object {
              "color": "white",
            }
          }
          testID="create-new-game-button-text"
        >
          Crear partida
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
